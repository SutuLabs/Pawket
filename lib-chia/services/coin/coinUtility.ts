import { Hex0x, prefix0x } from "./condition";
import { Bytes, bigint_to_bytes } from "clvm";
import { GetPuzzleApiCallback } from "../transfer/transfer";
import { CoinItem, CoinRecord } from "../../models/wallet";
import { CoinSpend, OriginCoin } from "../spendbundle";
import { AccountEntity } from "../../models/account";

export interface NetworkContext {
  prefix: string;
  symbol: string;
  chainId: string;
  api: GetPuzzleApiCallback;
}

export interface NetworkContextWithOptionalApi {
  prefix: string;
  symbol: string;
  chainId: string;
  api?: GetPuzzleApiCallback;
}

export interface CompatibleCoin {
  amount: bigint | number;
  parent_coin_info: Hex0x | string;
  puzzle_hash: Hex0x | string;
}

export interface LockedCoin {
  coinName: string;
  coin: OriginCoin;
  transactionTime: number;
  network: string;
  symbol?: string;
  accountFinger: number;
}

export interface PendingTransaction {
  coin: LockedCoin[];
  time: number;
  network: string;
  amount: { [key: string]: bigint };
}

export async function lockCoins(
  account: AccountEntity,
  coinSpends: CoinSpend[],
  transactionTime: number,
  network: string
): Promise<void> {
  const lc: LockedCoin[] = getLockedCoinsFromLocalStorage();
  const accountFinger = account.key.fingerprint;
  for (const cs of coinSpends) {
    const coinName = getCoinName(cs.coin);
    const idx = account.activities?.findIndex((act) => act.coin?.puzzleHash == cs.coin.puzzle_hash);
    if (idx != undefined && idx > -1)
      lc.push({
        coinName: coinName,
        coin: cs.coin,
        transactionTime: transactionTime,
        network: network,
        symbol: account.activities && account.activities[idx].symbol,
        accountFinger: accountFinger,
      });
  }
  localStorage.setItem("LOCKED_COINS", JSON.stringify(lc));
}

export function unlockCoins(coins: OriginCoin[]): OriginCoin[] {
  const lc: LockedCoin[] = getLockedCoinsFromLocalStorage();
  const uc: OriginCoin[] = [];
  for (const coin of coins) {
    if (!coin) continue;
    const cname = getCoinName(coin);
    const idx = lc.findIndex((c) => c.coinName == cname);
    if (idx > -1) {
      uc.push(coin);
      lc.splice(idx, 1);
    }
  }
  localStorage.setItem("LOCKED_COINS", JSON.stringify(lc));
  return uc;
}

export function getCompletedTransactions(coins: OriginCoin[], activities: CoinRecord[]): number[] {
  const txs: number[] = [];
  for (const sAct of activities) {
    if (sAct.spent && sAct.coin) {
      const idx = coins.findIndex(
        (c) => c.puzzle_hash == sAct.coin?.puzzleHash && c.parent_coin_info == sAct.coin.parentCoinInfo
      );
      if (idx > -1) txs.push(sAct.spentBlockIndex);
    }
  }
  return txs;
}

export function coinFilter(account: AccountEntity, coins: OriginCoin[], network: string): OriginCoin[] {
  let lc: LockedCoin[] = getLockedCoinsFromLocalStorage();
  lc = lc.filter((l) => l.network == network && l.accountFinger == account.key.fingerprint);
  return coins.filter((coin) => {
    const name = getCoinName(coin);
    return lc.findIndex((c) => c.coinName == name) == -1;
  });
}

export function getLockedCoinsFromLocalStorage(): LockedCoin[] {
  const lcStr = localStorage.getItem("LOCKED_COINS");
  let lc: LockedCoin[];
  try {
    lc = lcStr ? JSON.parse(lcStr) : [];
  } catch (error) {
    lc = [];
  }
  return lc;
}

export function getCoinName0x(coin: CompatibleCoin): Hex0x {
  return prefix0x(getCoinNameHex(coin).hex());
}

export function getCoinName(coin: CompatibleCoin): string;
export function getCoinName(coin: undefined): undefined;
export function getCoinName(coin: CompatibleCoin | undefined): string | undefined;
export function getCoinName(coin: CompatibleCoin | undefined): string | undefined {
  if (!coin) return coin;
  return getCoinNameHex(coin).hex();
}

export function getCoinNameHex(coin: CompatibleCoin): Bytes {
  const a = bigint_to_bytes(BigInt(coin.amount), { signed: true });
  const pci = Bytes.from(coin.parent_coin_info, "hex");
  const ph = Bytes.from(coin.puzzle_hash, "hex");
  const cont = pci.concat(ph).concat(a);
  const coinname = Bytes.SHA256(cont);
  return coinname;
}

export function convertToOriginCoin(
  coin: CoinItem | { amount: number; parent_coin_info: Hex0x | string; puzzle_hash: Hex0x | string }
): OriginCoin {
  return "parentCoinInfo" in coin
    ? {
        amount: BigInt(coin.amount),
        parent_coin_info: coin.parentCoinInfo,
        puzzle_hash: coin.puzzleHash,
      }
    : {
        amount: BigInt(coin.amount),
        parent_coin_info: prefix0x(coin.parent_coin_info),
        puzzle_hash: prefix0x(coin.puzzle_hash),
      };
}
