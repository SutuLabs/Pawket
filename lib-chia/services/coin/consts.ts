import { Bytes, bigint_from_bytes } from "clvm";

export const GROUP_ORDER = BigInt(
  bigint_from_bytes(Bytes.from("0x73EDA753299D7D483339D80809A1D80553BDA402FFFE5BFEFFFFFFFF00000001", "hex"))
);
export const DEFAULT_HIDDEN_PUZZLE_HASH = Bytes.from("711d6c4e32c92e53179b199484cf8c897542bc57f2b22582799f9d657eec4699", "hex");
