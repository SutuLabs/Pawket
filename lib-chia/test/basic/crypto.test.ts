import puzzle from "../../services/crypto/puzzle";
import utility from "../../services/crypto/utility";
import transfer from "../../services/transfer/transfer";
import { convertToOriginCoin, getCoinName } from "../../services/coin/coinUtility";
import { Instance } from "../../services/util/instance";
import { analyzeDidCoin } from "../../services/coin/did";
import { prefix0x } from "../../services/coin/condition";
import { analyzeNftCoin } from "../../services/coin/nft";
import { calculate_synthetic_secret_key, getSignMessage, signMessage, verifySignature } from "../../services/crypto/sign";
import { G2Element, PrivateKey } from "@chiamine/bls-signatures";

import didcoin2 from "../cases/didcoin2.json";
import nftcoin6 from "../cases/nftcoin6.json";
import { ByteBase, CryptographyService, EcPrivateKey, EcPublicKey } from "../../services/crypto/encryption";
import { EcdhHelper } from "../../services/crypto/ecdh";
import { getTestAccountWithPuzzles } from "../utility";
import { DEFAULT_HIDDEN_PUZZLE_HASH } from "../../services/coin/consts";

beforeAll(async () => {
  await Instance.init();
});

jest.setTimeout(30000);

test("Basic Cryptography", async () => {
  const hash = await utility.hash("test");
  expect(hash).toMatchSnapshot("hash");

  const privkey = utility.fromHexString("67b3dcf5ba985f77b7bb78b3edfd7e501f4669a3530b74f2247256e38b0529e2");
  const sk = await utility.getPrivateKey(privkey);
  const masterprikey = utility.toHexString(sk.serialize());
  const masterpubkey = utility.toHexString(sk.get_g1().serialize());
  expect(masterprikey).toMatchSnapshot("masterprikey");
  expect(masterpubkey).toMatchSnapshot("masterpubkey");
  expect(sk.get_g1().get_fingerprint().toString()).toMatchSnapshot("fingerprint");

  const derive = await utility.derive(privkey, true);
  const farmerpubkey = utility.toHexString(derive([12381, 8444, 0, 0]).get_g1().serialize());
  expect(farmerpubkey).toMatchSnapshot("farmerpubkey");
  const poolpubkey = utility.toHexString(derive([12381, 8444, 1, 0]).get_g1().serialize());
  expect(poolpubkey).toMatchSnapshot("poolpubkey");
  const walletprikey = utility.toHexString(derive([12381, 8444, 2, 0]).serialize());
  expect(walletprikey).toMatchSnapshot("walletprikey");
  const walletpubkey = utility.toHexString(derive([12381, 8444, 2, 0]).get_g1().serialize());
  expect(walletpubkey).toMatchSnapshot("walletpubkey");

  const adr = await puzzle.getAddress(walletpubkey, "xch");
  expect(adr).toMatchSnapshot("address");
});

test("Puzzle Assemble", async () => {
  const gt1 = transfer.getDelegatedPuzzle([
    ["a", "b", "c"],
    ["d", "e", "f"],
  ]);
  expect(gt1).toMatchSnapshot();

  const gt2 = transfer.getDelegatedPuzzle([
    ["a", "b", "c"],
    ["d", "e", ["f", `"g"`], []],
  ]);
  expect(gt2).toMatchSnapshot();
});

test("Coin Name", async () => {
  const coinname = getCoinName({
    amount: 1000000000n,
    parent_coin_info: "0xcd299604b459e5ff20da17627d684ea143fc1b5b4165166943729d2d24305de8",
    puzzle_hash: "0x484aaab0cda1b0149df9adeddb0a5d28976ad259ff6173f4488cbc68f095eb79",
  });
  expect(coinname).toMatchSnapshot("coinname");
});

// ## Signing Verification
//
// verification from:
// 1. pk -> (synthetic pk, p2 owner puzzle hash)
// 2. signature
// 3. message for signing
// 4. NFT/DID/XCH -> p2 owner puzzle hash

// verification process:
// 1. (pk -> synthetic pk, signature, message) -> signature correctness
// 2. (pk, NFT/DID/XCH) -> (pk -> synthetic pk -> p2 owner puzzle hash, NFT/DID/XCH -> p2 owner puzzle hash) -> linkability correctness

test("Sign Message By XCH", async () => {
  const puzzleHash = "0x7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10";

  const privkey = utility.fromHexString("55c335b84240f5a8c93b963e7ca5b868e0308974e09f751c7e5668964478008f");
  const requests = await puzzle.getPuzzleDetails(privkey, "xch", 0, 10);
  const sk = requests.find((_) => prefix0x(_.hash) == puzzleHash)?.privateKey;
  if (!sk) fail("cannot find sk");

  await signMessageTest(sk, "747769", puzzleHash);
});

test("Sign Message By XCH With Double Quote In Message", async () => {
  const puzzleHash = "0x7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10";

  const privkey = utility.fromHexString("55c335b84240f5a8c93b963e7ca5b868e0308974e09f751c7e5668964478008f");
  const requests = await puzzle.getPuzzleDetails(privkey, "xch", 0, 10);
  const sk = requests.find((_) => prefix0x(_.hash) == puzzleHash)?.privateKey;
  if (!sk) fail("cannot find sk");

  await signMessageTest(sk, '{ "hello": "world" }', puzzleHash);
});

test("Sign Message By DID", async () => {
  const analysis = await analyzeDidCoin(didcoin2.puzzle_reveal, "", convertToOriginCoin(didcoin2.coin), didcoin2.solution);
  if (!analysis) fail("cannot parse did coin");

  const privkey = utility.fromHexString("55c335b84240f5a8c93b963e7ca5b868e0308974e09f751c7e5668964478008f");
  const requests = await puzzle.getPuzzleDetails(privkey, "xch", 0, 10);
  const sk = requests.find((_) => prefix0x(_.hash) == analysis.hintPuzzle)?.privateKey;
  if (!sk) fail("cannot find sk");

  await signMessageTest(sk, "747769", analysis.hintPuzzle);
});

test("Sign Message By NFT", async () => {
  const analysis = await analyzeNftCoin(nftcoin6.puzzle_reveal, "", convertToOriginCoin(nftcoin6.coin), nftcoin6.solution);
  if (!analysis) fail("cannot parse nft coin");

  const privkey = utility.fromHexString("55c335b84240f5a8c93b963e7ca5b868e0308974e09f751c7e5668964478008f");
  const requests = await puzzle.getPuzzleDetails(privkey, "xch", 0, 10);
  const sk = requests.find((_) => _.hash == analysis.p2Owner)?.privateKey;
  if (!sk) fail("cannot find sk");

  await signMessageTest(sk, "747769", analysis.p2Owner);
});

async function signMessageTest(sk: PrivateKey, message: string, expectPuzzleHash: string): Promise<void> {
  const pk = utility.toHexString(sk.get_g1().serialize());

  const msg = await getSignMessage(message);
  expect(utility.toHexString(msg)).toMatchSnapshot("msg");

  const { signature, syntheticPublicKey } = signMessage(sk, msg);
  expect(signature).toMatchSnapshot("signature");
  expect(syntheticPublicKey).toMatchSnapshot("syntheticPublicKey");

  const v = verifySignature(pk, msg, signature);
  expect(v).toBeTruthy();

  const vp2owner = prefix0x(await puzzle.getPuzzleHash(pk));
  expect(vp2owner).toBe(prefix0x(expectPuzzleHash));
}

test("ECDH 1", async () => {
  await testEncryption("hello");
});

test("ECDH 2", async () => {
  await testEncryption(`
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
  `);
});

async function testEncryption(plaintext: string): Promise<void> {
  const ecc = new CryptographyService();

  const sk1 = EcPrivateKey.parse("1d1b65531bf3b0fc629fb4537b8bf4083bb72b532f1b571250b137ca6bb4dde3");
  const sk2 = EcPrivateKey.parse("3d0578e760b3abb55eee2a207f86610f49c909b0bed5f1775e7c2cc1efe0033b");
  if (!sk1 || !sk2) fail();

  const pk1 = ecc.getPublicKey(sk1);
  const pk2 = ecc.getPublicKey(sk2);
  expect(pk1.toHex()).toMatchSnapshot("pk1");
  expect(pk2.toHex()).toMatchSnapshot("pk2");

  const pk1Parse = EcPublicKey.parse(pk1.toHex());
  expect(pk1Parse).toStrictEqual(pk1);

  const ran = ByteBase.hexStringToByte("3d0578e760b3abb55eee2a207f86610f");

  const enc = await ecc.encrypt(plaintext, pk2, sk1, ran);
  const dec = await ecc.decrypt(enc, pk1, sk2);
  expect(dec).toBe(plaintext);
  expect(enc).toMatchSnapshot("encrypted");
  expect(dec).toMatchSnapshot("decrypted");
}

test("BLS Aggregation", async () => {
  const BLS = Instance.BLS;
  if (!BLS) throw new Error("BLS not initialized");

  const sigs = [
    "c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "ac217e250273a881566563ffa4e296f7ed24ff6c44289b732875b67a0306766fd0e0675ec22cef94e5a7d349aba39ee103b33249be47c33a37698a5000fdda59b0bb39df89f74e6a457002ab5da32cafe24f7987e7722331736a9e2622c179d5",
  ].map((_) => BLS.G2Element.from_bytes(utility.fromHexString(_)));

  const serialize = (sig: G2Element | undefined): string => {
    if (!sig) return "";
    return utility.toHexString(sig.serialize());
  };
  const a1 = BLS.AugSchemeMPL.aggregate(sigs.slice(0, 1));
  const a2 = BLS.AugSchemeMPL.aggregate(sigs.slice(0, 2));
  const a3 = BLS.AugSchemeMPL.aggregate(sigs.slice(1, 2));
  const a4 = BLS.AugSchemeMPL.aggregate([a1, a3]);
  const a5 = BLS.AugSchemeMPL.aggregate([a1, a2]);
  const a8 = BLS.AugSchemeMPL.aggregate([a3]);
  expect(serialize(a1)).toBe(serialize(sigs.at(0)));
  expect(serialize(a2)).toBe(serialize(sigs.at(1)));
  expect(serialize(a3)).toBe(serialize(a2));
  expect(serialize(a4)).toBe(serialize(a2));
  expect(serialize(a5)).toBe(serialize(a2));
  expect(serialize(a8)).toBe(serialize(a2));

  const a6 = BLS.AugSchemeMPL.aggregate([a3, a2]);
  const a7 = BLS.AugSchemeMPL.aggregate([a2, a3]);

  expect(serialize(a6)).not.toBe(serialize(a2));
  expect(serialize(a6)).toMatchSnapshot();
  expect(serialize(a7)).toBe(serialize(a6));
});

test("BLS calculate_synthetic_secret_key", async () => {
  const ecdh = new EcdhHelper();
  const sk = "55c335b84240f5a8c93b963e7ca5b868e0308974e09f751c7e5668964478008f";

  const BLS = Instance.BLS;
  if (!BLS) throw new Error("BLS not initialized");

  const synsk_noble_bls = ecdh.calculate_synthetic_secret_key(utility.fromHexString(sk), DEFAULT_HIDDEN_PUZZLE_HASH.raw());
  const synsk_clvm_bls = calculate_synthetic_secret_key(
    BLS,
    BLS.PrivateKey.from_bytes(utility.fromHexString(sk), true),
    DEFAULT_HIDDEN_PUZZLE_HASH.raw()
  ).serialize();

  expect(utility.toHexString(synsk_noble_bls)).toBe(utility.toHexString(synsk_clvm_bls));
});

test("BLS ECDH 1", async () => {
  await testBlsEcdh("hello");
});

test("BLS ECDH 2", async () => {
  await testBlsEcdh(`
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
  `);
});

async function testBlsEcdh(plaintext: string): Promise<void> {
  const ecdh = new EcdhHelper();

  const account1 = await getTestAccountWithPuzzles("55c335b84240f5a8c93b963e7ca5b868e0308974e09f751c7e5668964478008f");
  const account2 = await getTestAccountWithPuzzles("46815978e90da660427161c265b400831ee59f9aae9a40b449fbcd67ca140590");
  const ph1 = prefix0x(account1.addressPuzzles[0].puzzles[1].hash);
  const ph2 = prefix0x(account2.addressPuzzles[0].puzzles[1].hash);
  global.fetch = jest.fn().mockImplementation(mockFetch);
  const random = utility.fromHexString("b449fbcd67ca14059046815978e90da6");

  const enc = await ecdh.encrypt(ph1, ph2, plaintext, account1, "", random);
  expect(enc).toMatchSnapshot("encrypted");

  const dec = await ecdh.decrypt(ph1, ph2, enc, account2, "");

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (<any>global.fetch).mockClear();

  expect(dec).toBe(plaintext);
}

function mockFetch(url: RequestInfo, args: RequestInit): Promise<Response> {
  return Promise.resolve(<never>{
    ok: true,
    status: 200,
    json: async () => {
      if (
        url == "Wallet/records" &&
        args.body ==
          '{"puzzleHashes":["0x83baea4313afa1a5b174d1afef81ac640688a3a0157c418536c06e2b556adc55"],"includeSpentCoins":true,"hint":false}'
      )
        return {
          coins: [
            {
              puzzleHash: "0x83baea4313afa1a5b174d1afef81ac640688a3a0157c418536c06e2b556adc55",
              records: [
                {
                  coin: {
                    parentCoinInfo: "0xdfba5fd91f26eda808f74bdf8a753454c6f1e13a9bd9b1bc8040be64ec510e24",
                    puzzleHash: "0x83baea4313afa1a5b174d1afef81ac640688a3a0157c418536c06e2b556adc55",
                    amount: 100,
                  },
                  spent: true,
                },
              ],
            },
          ],
        };

      if (
        url == "Wallet/records" &&
        args.body ==
          '{"puzzleHashes":["0x7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10"],"includeSpentCoins":true,"hint":false}'
      )
        return {
          coins: [
            {
              puzzleHash: "0x7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10",
              records: [
                {
                  coin: {
                    parentCoinInfo: "0x0068185e4f3a4217da08f1a04173f970d7f7f581e43f270ecc6c7260dd4bc38f",
                    puzzleHash: "0x7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10",
                    amount: 18,
                  },
                  spent: true,
                },
              ],
            },
          ],
        };

      if (
        url == "Wallet/get-puzzle" &&
        args.body == '{"parentCoinId":"0xc20ab64698a0fc4dc1583b984c97eecb9f9348fb3343bfc5f5c94ee50bc05956"}'
      )
        return {
          parentCoinId: "0xc20ab64698a0fc4dc1583b984c97eecb9f9348fb3343bfc5f5c94ee50bc05956",
          amount: 100,
          parentParentCoinId: "0xdfba5fd91f26eda808f74bdf8a753454c6f1e13a9bd9b1bc8040be64ec510e24",
          puzzleReveal:
            "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b0ae72dd3ae45995713c05bb674ebbfce59e68fdf7102b06cd6d38ebed11c12fe9bdec89ee4887e39382a4007f1d82b1f8ff018080",
        };

      if (
        url == "Wallet/get-puzzle" &&
        args.body == '{"parentCoinId":"0x1718f1cc4aa62ace337f04208e62cbce178a3c0edec9f85638cf7c093129c450"}'
      )
        return {
          parentCoinId: "0x1718f1cc4aa62ace337f04208e62cbce178a3c0edec9f85638cf7c093129c450",
          amount: 18,
          parentParentCoinId: "0x0068185e4f3a4217da08f1a04173f970d7f7f581e43f270ecc6c7260dd4bc38f",
          puzzleReveal:
            "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b092c7027d455e843b1226b757c1dd5e124e0e6ddab3549036a4684b410182fdfd1eee459eb76a4d6d73e94f90fbe6eb1dff018080",
        };

      console.log("unexpect fetch url and args", url, args);
      return {};
    },
  });
}
