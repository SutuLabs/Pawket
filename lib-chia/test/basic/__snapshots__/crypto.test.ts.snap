// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BLS Aggregation 1`] = `"b48d90ebddc4fafc50d68fd0b6902583479c1ec77788d6a2dfaae1c9b4559b4909f0fa9933d475711a79a0e8c7e562ea0d406d8caa53a97f0b40ea2a37a5eab15ccdad6f1c8f1b4bd82302ef07b2037d54ee9ab7a22cc0938ba3e329e5b445b7"`;

exports[`BLS ECDH 1: encrypted 1`] = `"b449fbcd67ca14059046815978e90da60cea883805cbe120146375dde3295ca0"`;

exports[`BLS ECDH 2: encrypted 1`] = `"b449fbcd67ca14059046815978e90da6172d60ae943c08611d65d7195fcf617e3fd8fd806a2f3f5045d0f1040a503722dbedbab09d753fd82d1fae4c7bff564f718cde37d17580891d30bf930d0f13d6b9dafa300f484f7d68fb06d7ca93e45a5b3bc1d978ffe739adfebafe347bda11029c2dd9c48ee0d4461b816e716a9376daa9065c13578483896bbe033d9e4dd65bce230b21d4ff7cb00e08b80ab925c95519537d150ae760366cd06ffec2fac12c76faeeacfd288f2cba949fc07d7f45eabdde2c24b53858f002282bce4a784b053f3312e4579223fa45b782356a6920fdbc160f8cc7827d0b39beac1716a1516e65c5c1ff0bbcff27c3736a2da1f2ddae4a8168c0765d426913e909b143d0b68c8a5a47e31e0ac47648daa3f5f28522"`;

exports[`Basic Cryptography: address 1`] = `"xch13akv0y3er0qvdjwzks2gm4ljj7qpynrh6rcsnwc6y0hyfgzdj89sr43zcp"`;

exports[`Basic Cryptography: farmerpubkey 1`] = `"99f181b1b30865339eea4e61172e64282ed1793022363f6635affbfb2aa2f4384ef10df47a445f92d685c7265488777e"`;

exports[`Basic Cryptography: fingerprint 1`] = `"2308828858"`;

exports[`Basic Cryptography: hash 1`] = `"9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08"`;

exports[`Basic Cryptography: masterprikey 1`] = `"67b3dcf5ba985f77b7bb78b3edfd7e501f4669a3530b74f2247256e38b0529e2"`;

exports[`Basic Cryptography: masterpubkey 1`] = `"89d4ca795881dc192cf7e0bcf22528c3d7708f57aa644647ae2506cd10405dc528b36e450ab901675f7a6289a4335a22"`;

exports[`Basic Cryptography: poolpubkey 1`] = `"9896eb01246db07360b4caba1861a6d13c5be923e28a68ba8429d841c5dfa429e05ef639003793c4715a8be4578abe9e"`;

exports[`Basic Cryptography: walletprikey 1`] = `"0408d4c5e7c97af49dd5605db1907a231761c0cba816c778e74c8de24d0793cd"`;

exports[`Basic Cryptography: walletpubkey 1`] = `"8b58921998b2337fd9d4a410e8bd11dfda435feb92dc3d6f3e111194b99a0391f39cf9b9534a44d154c9d7492400f36b"`;

exports[`Coin Name: coinname 1`] = `"c62152493a6f3fbca7a918a8258e52f84c4e8aa286b8ebf1b0d5e5dda7fa7e6f"`;

exports[`ECDH 1: decrypted 1`] = `"hello"`;

exports[`ECDH 1: encrypted 1`] = `"3d0578e760b3abb55eee2a207f86610f6979d828e7f013c3ab07be3512673944"`;

exports[`ECDH 1: pk1 1`] = `"72a493cf43d9a55d245453eb672aa2e52b8f7cf0bcf5bb16bdbf7f7187a002dc"`;

exports[`ECDH 1: pk2 1`] = `"74c7c8b01b67448c6199d848ac16884b50216e7266ec4fa21f9c5d6b0c9ffdab"`;

exports[`ECDH 2: decrypted 1`] = `
"
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
very long sentence with newline
  "
`;

exports[`ECDH 2: encrypted 1`] = `"3d0578e760b3abb55eee2a207f86610f3d438ba728ec2d91b98ded328fc922c60521dde55f7123982fa14bd80c051e5ba4d384a3023e171106630069ff6f38d3b31aebee6f898f6ab21bc99e2af1c2b60a8d30490a71da5f349a64485eafccb70ff1042d5b9c29ad75c97c6e18529228fb6357831cb16fdfff6c45989f87d93b4f3450fcb5995de197b5fd057f223718069955323de8b88b2a296b244f82995e9693ce1808a8c2522596bb81e058e41faf867e816a3aede95f4b6c78f1f69e837687eb1d6e1a4ba598f05a1fd85d1ec434cb21bfa998de63665764ad4000376537525e0b428df13abdf45061af69024a1983312a798017e5d7bc31ef7173195610f1b5b54ec85f79f29f85c92ea60cd6c672dfb7fcb95bbe4d4fb43aea263991"`;

exports[`ECDH 2: pk1 1`] = `"72a493cf43d9a55d245453eb672aa2e52b8f7cf0bcf5bb16bdbf7f7187a002dc"`;

exports[`ECDH 2: pk2 1`] = `"74c7c8b01b67448c6199d848ac16884b50216e7266ec4fa21f9c5d6b0c9ffdab"`;

exports[`Puzzle Assemble 1`] = `"(q (a b c) (d e f))"`;

exports[`Puzzle Assemble 2`] = `"(q (a b c) (d e (f \\"g\\") ()))"`;

exports[`Sign Message By DID: msg 1`] = `"eac753f6f07152ef558a6fd1d1b989cc99f0528bb34f5b927294f031117b3d70"`;

exports[`Sign Message By DID: signature 1`] = `"0x82ca9a652a58e148ec74645dae3106c28f33d5057abedec9e7fbe6b9d46450dd71082bb5dbff4108a75184f7825d7857053da2caa2bafb12e8f689ff6f24e88f930d515244d9730300c860ac5279ae8cca2022ea0588b6b68f25e6fc80471197"`;

exports[`Sign Message By DID: syntheticPublicKey 1`] = `"0xb79284e00045f07be014cacf9aac17b5a41648b32329eb2a38104768a0a0abd6542cb577b28e25458a8472c43fb8475b"`;

exports[`Sign Message By NFT: msg 1`] = `"eac753f6f07152ef558a6fd1d1b989cc99f0528bb34f5b927294f031117b3d70"`;

exports[`Sign Message By NFT: signature 1`] = `"0x8310a5ce30132bd240595fa9da662bbaa77b505feedfb19fdafb121113c354bd3ecfad0d46851816fd0ad3cb6ea6a05e0506f44924cd4d7c0844f1ac54c8355e0194eec826784fd4bc458a409b53ca060beacf4dd06d309ffaeee31e1faf22c0"`;

exports[`Sign Message By NFT: syntheticPublicKey 1`] = `"0xa0ea9df64dd859d6d7e648fec94b35ee0512670e68c0a83c520948dc164d817d39c402002db1cf716caa1335022c8338"`;

exports[`Sign Message By XCH With Double Quote In Message: msg 1`] = `"fcb527aec477a179db84da00ca1b9e0ba55a2a38117bcca9a0e3656ce5b87124"`;

exports[`Sign Message By XCH With Double Quote In Message: signature 1`] = `"0x8e5feb29753cae76b2504c8c24595c0d533f057b53476642859020e8d02604c9d7584d2cfc819585f5b43a8129083ea210c6dc1189c44140a7defb539f43502b42dda93a00c11abe5e339ce0c74f5ff16df6615380e3358935b6753badc59167"`;

exports[`Sign Message By XCH With Double Quote In Message: syntheticPublicKey 1`] = `"0x92c7027d455e843b1226b757c1dd5e124e0e6ddab3549036a4684b410182fdfd1eee459eb76a4d6d73e94f90fbe6eb1d"`;

exports[`Sign Message By XCH: msg 1`] = `"eac753f6f07152ef558a6fd1d1b989cc99f0528bb34f5b927294f031117b3d70"`;

exports[`Sign Message By XCH: signature 1`] = `"0xa673160a951c23683f4d8aa551eca97767befbf025d75f4539e7db4aa639efe950f389134e5c2552f5e2bd87c3ce97600d523cb434cdf75afe6a490b4dba4cb2bba5e7c49b44c55743bffbe08b372955d74fd0384178ebf40ee73a1b46d27f8f"`;

exports[`Sign Message By XCH: syntheticPublicKey 1`] = `"0x92c7027d455e843b1226b757c1dd5e124e0e6ddab3549036a4684b410182fdfd1eee459eb76a4d6d73e94f90fbe6eb1d"`;
